clients/pkg/promtail/scrapeconfig/scrapeconfig.go

```diff
@@ -159,6 +159,9 @@ type JournalTargetConfig struct {
 	// Path to a directory to read journal entries from. Defaults to system path
 	// if empty.
 	Path string `yaml:"path"`
+
+	// Journal matches to filter. Disjunctions (+) not supported.
+	Matches string `yaml:"matches"`
```

Comment:
I though about it before submitted this PR, but if "github.com/coreos/go-systemd/sdjournal" support logical AND (some work can be seen in their source code) it looked a bit weird to me to add a single + character as a list item.

So finally I decided to go with a string and the idea to set here the same arguments that you can use with journalctl, for example "_TRANSPORT=kernel + PRIORITY=3", but I don't have a strong opinion which two options is better.

Reply:
Thanks for the detail :+1: 
I think we shouldn't let the implementation detail of a library we use bubble up to the end-user, so perhaps we can specify the matches as a list of conditions (I think this is the most ergonomic / idiomatic "YAML way"), and then provide them to the library in the way it requires.

Reply:
I think the problem is precisely the implementation detail of the library that promtail uses. It uses and array of matches and this decision limits the way you could set AND/OR matches. Maybe that's the cause that this library doesn't implement it at all taking into account that Journal supports AND/OR matches, and the library is just a binding of journald code.

Journal support filters with logical AND and/or logical OR, so I just though how to support that expression of filters in promtail without bringing a no ideal (IMO) decision of the library to promtail.

Anyway, the change proposed is quite simple and could be easily coded although I think it should not be done.

Reply:
OK, i've taken a look at the [manpage](https://man7.org/linux/man-pages/man1/journalctl.1.html) to re-familiarise myself with the specifics of filtering.

Here's one example they cite:

>        To show all fields emitted by a unit and about the unit, option
>        -u/--unit= should be used.  journalctl -u name expands to a
>        complex filter similar to
> 
>            _SYSTEMD_UNIT=name.service
>              + UNIT=name.service _PID=1
>              + OBJECT_SYSTEMD_UNIT=name.service _UID=0
>              + COREDUMP_UNIT=name.service _UID=0 MESSAGE_ID=fc2e22bc6ee647b6b90729ab34a250b1

If a user were to try write a filter expression like the above in a promtail config, might it look like this?

```yaml
  ...
  matches:
    - _SYSTEMD_UNIT=name.service
    - UNIT=name.service _PID=1
    - OBJECT_SYSTEMD_UNIT=name.service _UID=0
    - COREDUMP_UNIT=name.service _UID=0 MESSAGE_ID=fc2e22bc6ee647b6b90729ab34a250b1
  ...
```

In the documentation we could specify that each `matches` item is a disjunction (logical `OR`), and the logical `AND` is combining multiple filters in a single entry separated by space.

It seems the library [supports adding disjunctions](https://github.com/coreos/go-systemd/blob/main/sdjournal/journal.go#L524), but I'm not familiar enough with the code to know how much effort this will be.

I understand that you've put in a lot of effort into this already, so I'm happy to merge this as-is, but in a perfect world I think we should follow the [Principle of Least Surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) here to make the filtering work as a user would expect. If journal filtering allows for disjunctions, then we should support it, too.

If you don't have the time to add disjunction support, that's ok - we can always add it as an enhancement for later; I don't want to block this useful feature.

Reply:
> In the documentation we could specify that each matches item is a disjunction (logical OR), and the logical AND is combining multiple filters in a single entry separated by space.

 Absolutely. Logical AND have preference and are evaluated first, so IMO that's the way to express the filters. Or going further in the "yaml way" a two levels nested list:

```
  ...
  matches:
    - _SYSTEMD_UNIT=name.service
    - - UNIT=name.service
      - _PID=1
    - - OBJECT_SYSTEMD_UNIT=name.service
      - _UID=0
    - - COREDUMP_UNIT=name.service
      - _UID=0 
      - MESSAGE_ID=fc2e22bc6ee647b6b90729ab34a250b1
  ...
```
In the short time, in order to user current go-systemd functionality there should be only one list item:
```
  ...
  matches:
    - COREDUMP_UNIT=name.service _UID=0 MESSAGE_ID=fc2e22bc6ee647b6b90729ab34a250b1
  ...
```
or
```
  ...
  matches:
    - - COREDUMP_UNIT=name.service
      - _UID=0 
      - MESSAGE_ID=fc2e22bc6ee647b6b90729ab34a250b1
  ...

```
which is a bit weird and we must explain it in the docs. So we have a trade off between being ready to support future functionality and a weird (IMO) configuration or have a more simple configuration and no-compatibility with future functionality.

> It seems the library [supports adding disjunctions](https://github.com/coreos/go-systemd/blob/main/sdjournal/journal.go#L524), but I'm not familiar enough with the code to know how much effort this will be.

Yes, it is ready to use disjuntions, but the [reader](https://github.com/coreos/go-systemd/blob/main/sdjournal/read.go#L67) does not uses them at all. IMO the library should mimic [journalctl add_matches](https://github.com/systemd/systemd/blob/main/src/journal/journalctl.c#L1092). 

My main concern here is that proposing a modification in coreos/go-systemd reader to support disjunctions will need a different JournalReaderConfig struct, maybe replacing the match list with a two levels nested list of matches, and this will break compatibility of anyone using it, like promtail.

I will reach you in your public slack so we can talk about next steps.

Reply:
@carlospeon heads up, I'm on PTO from tomorrow and back on the 22nd Aug.
I'm happy for this to be merged as-is once all the comments are addressed (except for this one)